src/Compilers/CSharp/Portable/BoundTree/Expression.cs

```diff
@@ -0,0 +1,1414 @@
+        {
+            int index = ArgumentIndexMatchingParameter(arguments, argumentsToParameters, parameter.ContainingSymbol as IMethodSymbol, parameter);
+            if (index >= 0)
+            {
+                return DeriveArgument(index, arguments, argumentNames, argumentRefKinds, targetMethod.Parameters.Length, targetMethod.Parameters[targetMethod.Parameters.Length - 1].IsParams);
+            }
+
+            return null;
+        }
+
+        static int ArgumentIndexMatchingParameter(ImmutableArray<BoundExpression> arguments, ImmutableArray<int> argumentsToParameters, IMethodSymbol targetMethod, IParameterSymbol parameter)
+        {
+            if (parameter.ContainingSymbol == targetMethod)
+            {
+                int parameterIndex = parameter.Ordinal;
+                ImmutableArray<int> parameterIndices = argumentsToParameters;
+                if (!parameterIndices.IsDefaultOrEmpty)
+                {
+                    for (int index = 0; index < parameterIndices.Length; index++)
+                    {
+                        if (parameterIndices[index] == parameterIndex)
+                        {
+                            return index;
+                        }
+                    }
+                }
+
+                return parameterIndex;
+            }
+
+            return -1;
+        }
+
+        class SimpleArgument : IArgument
+        {
+            readonly IExpression ArgumentValue;
+            public SimpleArgument(IExpression value)
+            {
+                this.ArgumentValue = value;
+            }
+
+            public ArgumentKind Kind => ArgumentKind.Positional;
+            
+            public ArgumentMode Mode => ArgumentMode.In;
+           
+            public IExpression Value => this.ArgumentValue;
+
+            public IExpression InConversion => null;
+
+            public IExpression OutConversion => null;
+        }
+
+        class Argument : IArgument
+        {
+            readonly IExpression ArgumentValue;
+            readonly ArgumentKind ArgumentKind;
+            readonly ArgumentMode ArgumentMode;
+            public Argument(ArgumentKind kind, ArgumentMode mode, IExpression value)
+            {
+                this.ArgumentValue = value;
+                this.ArgumentKind = kind;
+                this.ArgumentMode = mode;
+            }
+
+            public ArgumentKind Kind => this.ArgumentKind;
+           
+            public ArgumentMode Mode => this.ArgumentMode;
+            
+            public IExpression Value => this.ArgumentValue;
+
+            public IExpression InConversion => null;
+
+            public IExpression OutConversion => null;
+        }
+
+        class NamedArgument : Argument, INamedArgument
```

Comment:
I feel like there is value in these classes.  But i'm less convinced there is value in the different interfaces here.